src/wh_server_crypto.c

```diff
@@ -649,6 +647,79 @@ int wh_Server_HandleCryptoRequest(whServerContext* server,
         }
         break;
 #endif /* !WC_NO_RNG */
+#ifdef WOLFSSL_CMAC
+    case WC_ALGO_TYPE_CMAC:
+        /* in, out and key are after the fixed size fields */
+        in = (uint8_t*)(&packet->cmacReq + 1);
+        key = in + packet->cmacReq.inSz;
+        out = (uint8_t*)(&packet->cmacRes + 1);
+        /* do oneshot if all fields are present */
+        if (packet->cmacReq.inSz != 0 && packet->cmacReq.keySz != 0 &&
+            packet->cmacReq.outSz != 0) {
+            field = packet->cmacReq.outSz;
+            ret = wc_AesCmacGenerate_ex(server->crypto->cmac, out, &field, in,
+                packet->cmacReq.inSz, key, packet->cmacReq.keySz, NULL,
+                server->crypto->devId);
+            packet->cmacRes.outSz = field;
+        }
+        else {
+            /* do each operation based on which fields are set */
+            /* set key if present, otherwise load the struct from keyId */
+            if (packet->cmacReq.keySz != 0) {
+                /* initialize cmac with key and type */
+                ret = wc_InitCmac_ex(server->crypto->cmac, key,
+                    packet->cmacReq.keySz, packet->cmacReq.type, NULL, NULL,
+                    server->crypto->devId);
+            }
+            else {
+                field = sizeof(server->crypto->cmac);
+                ret = hsmReadKey(server,
+                    MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_CRYPTO,
+                    server->comm->client_id, packet->cmacReq.keyId), NULL,
+                    (uint8_t*)server->crypto->cmac, &field);
+            }
+            if (ret == 0 && packet->cmacReq.inSz != 0) {
+                ret = wc_CmacUpdate(server->crypto->cmac, in,
+                    packet->cmacReq.inSz);
```

Comment:
Need to make sure keyId hasn't changed.

Reply:
If keyId changes and it's a non-existent key hsmReadKey will return not found. If it's a false key I could check for size to make sure it's the same size as the cmac struct but it's up to the client to send the right keyId for any operation and the cryptocb sets devCtx so as long as the caller isn't messing with it it will be the same